File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    # DFS
    def dfs(depth, visited, nums):
        nonlocal dungeons, answer, k
        # 탐색 끝
        if depth == len(dungeons):
            hp = k
            cnt = 0
            for i in nums:
                if dungeons[i][0] <= hp:
                    hp -= dungeons[i][1]
                    cnt += 1
                else:
                    break
            answer = max(answer, cnt)
            return
        # 탐색
        for i in range(len(dungeons)):
            if visited[i]:
                continue
            nums.append(i)
            visited[i] = True
            dfs(depth+1, visited, nums)
            visited[i] = False
            nums.pop()
    
    # Main
    answer = 0
    visited = [False]*len(dungeons)
    dfs(0, visited, [])
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.io.*;
import java.util.*;

class Solution {
    
    int result = -1;
    int totalSize = -1;
    
    public int backTracking(boolean[] visited, int explored, int remains, int[][] dungeons) {
        
        // 결과 업데이트
        result = Math.max(result, explored);
        
        // 순회하면서 방문할 수 있는지 보기
        for ( int i = 0 ; i < totalSize ; i++ ) {
            
            // 이미 방문했다면 넘기기
            if ( visited[i] ) {
                continue;
            }
            
            // 최소 피로도 조건 안되면 넘기기
            if ( dungeons[i][0] > remains ) {
                continue;
            }
            
            // 백트래킹
            visited[i] = true;
            backTracking(visited, explored + 1, remains - dungeons[i][1], dungeons);
            visited[i] = false;
            
        }
        
        return 0;
        
    }
    
    public int solution(int k, int[][] dungeons) {
        
        // 초기화
        totalSize = dungeons.length;
        boolean[] visited = new boolean[totalSize];
        
        // 백트래킹
        backTracking(visited, 0, k, dungeons);
        return result;
    }
    
}